docs/reference/troubleshooting/corruption-issues.asciidoc

```diff
@@ -0,0 +1,82 @@
+during merges, shard movements, and snapshots. This does not mean that these
+proceses are causing corruption: they are examples of the rare times where
+reading a whole file is necessary. {es} takes the opportunity to verify the
+checksum at the same time, and this is when the corruption is detected and
+reported. It doesn't indicate the cause of the corruption or when it happened.
+Corruptions can remain undetected for many months.
+
+The files that make up a Lucene index are written sequentially from start to
+end and then never modified or overwritten. This access pattern means the
+checksum computation is very simple and can happen on-the-fly as the file is
+written, and also makes it very unlikely that an incorrect checksum is due to a
+userspace bug at the time the file was written. The routine that computes the
+checksum is straightforward, widely used, and very well-tested, so you can be
+very confident that a checksum mismatch really does indicate that the data read
+from disk is different from the data that {es} previously wrote.
+
+The files that make up a Lucene index are written in full before they are used.
+If a file is needed to recover an index after a restart then your storage
+system will previously have confirmed to {es} that this file was durably synced
+to disk. On Linux this means that the `fsync()` system call returned
+successfully. {es} sometimes detects that an index is corrupt because a file
+needed for recovery has been truncated or is missing its footer. This indicates
+that your storage system acknowledges durable writes incorrectly.
+
+There are many possible explanations for {es} detecting corruption in your
+cluster. Databases like {es} generate a challenging I/O workload that may find
+subtle infrastructural problems which other tests may miss. {es} is known to
+expose the following problems as file corruptions:
+
+- filesystem bugs, especially in newer and nonstandard filesystems which might
+  not have seen enough real-world production usage to be confident that they
+work correctly.
+
+- https://www.elastic.co/blog/canonical-elastic-and-google-team-up-to-prevent-data-corruption-in-linux[kernel bugs]
+
+- bugs in firmware running on the drive or RAID controller
+
+- incorrect configuration, for instance configuring `fsync()` to report success
+  before all durable writes have completed
+
+- faulty hardware (for instance the drive, RAID controller, RAM or CPU)
+
+Data corruption typically doesn't result in other evidence of problems apart
+from the checksum mismatch. Do not interpret this as an indication that your
+storage subsystem is working correctly and therefore that {es} itself caused
+the corruption. It is rare for faulty storage to show any evidence of problems
+apart from the data corruption, but data corruption itself is a very strong
+indicator that your storage subsystem is not working correctly.
+
+To narrow down the source of the corruptions, systematically change components
```

Comment:
I think I'd like to reword this a bit to be more of a non-exhaustive list of suggestion of ways to help diagnosing corruptions.

Reply:
:+1: see 37a0caa4d.